filters/include/pcl/filters/impl/spherical_voxel_grid.hpp

```diff
@@ -0,0 +1,318 @@
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id$
+ *
+ */
+
+#ifndef PCL_FILTERS_SPHERICAL_VOXEL_GRID_IMPL_H_
+#define PCL_FILTERS_SPHERICAL_VOXEL_GRID_IMPL_H_
+
+#include <pcl/filters/spherical_voxel_grid.h>
+#include <pcl/filters/voxel_grid.h>
+#include <pcl/common/common.h>
+#include <pcl/common/centroid.h>
+
+struct cloud_point_index_idx
+{
+  unsigned int idx;
+  unsigned int cloud_point_index;
+
+  cloud_point_index_idx (unsigned int idx_, unsigned int cloud_point_index_) : idx (idx_), cloud_point_index (cloud_point_index_) {}
+  bool operator < (const cloud_point_index_idx &p) const { return (idx < p.idx); }
+};
+
+template <typename PointT> void
+pcl::SphericalVoxelGrid<PointT>::applyFilter (PointCloud &output)
+{
+  // Has the input dataset been set already?
+  if (!input_)
+  {
+    PCL_WARN ("[pcl::%s::applyFilter] No input dataset given!\n", getClassName().c_str());
+    output.width = output.height = 0;
+    output.points.clear();
+    return;
+  }
+
+  // Copy the header (and thus the frame_id) + allocate enough space for points
+  output.height = 1;                    // downsampling breaks the organized structure
+  output.is_dense = true;                 // we filter out invalid points
+
+  if (leaf_size_r_ <= 0 || leaf_size_theta_ <= 0 || leaf_size_phi_ <= 0)
+  {
+    PCL_WARN("[pcl::%s::applyFilter] Leaf size improperly set.", getClassName().c_str());
+    output = *input_;
+    return;
+  }
+
+  // Find the farthest away point, subject to filter parameters if specified
+  Eigen::Vector4f max_p;
+
+  if (!filter_field_name_.empty ()) // If we don't want to process the entire cloud...
+  {
+    std::vector<int> indices_filtered;
+    indices_filtered.reserve(indices_->size());
+
+    // Get the distance field index
+    std::vector<pcl::PCLPointField> fields;
+    int filter_idx = pcl::getFieldIndex (*input_, filter_field_name_, fields);
+    if (filter_idx == -1)
+      PCL_WARN ("[pcl::%s::applyFilter] Invalid filter field name. Index is %d.\n", getClassName ().c_str (), filter_idx);
+
+    for (std::vector<int>::const_iterator it = indices_->begin (); it != indices_->end (); ++it)
+    {
+      if (!input_->is_dense)
+        // Check if the point is invalid
+        if (!pcl_isfinite (input_->points[*it].x) ||
+            !pcl_isfinite (input_->points[*it].y) ||
+            !pcl_isfinite (input_->points[*it].z))
+          continue;
+
+      // Get the filter value
+      const uint8_t* pt_data = reinterpret_cast<const uint8_t*> (&input_->points[*it]);
+      float filter_value = 0;
+      memcpy (&filter_value, pt_data + fields[filter_idx].offset, sizeof (float));
```

Comment:
Being picky, not every field is of type `float`. Intended use is with `float` fields though, so this is probably ok.

Reply:
How about I introduce a check to verify that the field is a float?